File: 백준/Gold/9663. N－Queen/N－Queen.py


def queen(v):
    for i in range(v):
        if visited[v] == visited[i] or abs(visited[v] - visited[i]) == abs(v - i):
            return False
    return True


def dfs(v):
    global cnt
    if v == N:
        cnt += 1
    else:
        for i in range(N):
            visited[v] = i
            if queen(v):
                dfs(v+1)


N = int(input())
cnt = 0
visited = [0] * N
dfs(0)

print(cnt)

File: 백준/Gold/9663. N－Queen/N－Queen.cc
#include <iostream>

using namespace std;

int N, ans;
int visited[15];

bool flag(int v)
{
    for (int i = 0; i < v; i++)
        if (visited[v] == visited[i] || abs(visited[v] - visited[i]) == abs(v-i))
            return 0;
    
    return 1;
}

void dfs(int depth)
{
    if (depth == N)
    {
        ans++;
        return;
    }

    for (int i = 0; i < N; i++)
    {
        visited[depth] = i;

        if (flag(depth))
            dfs(depth+1);
    }
}

int main()
{
    ios_base::sync_with_stdio(false);
    cin.tie(NULL);

    cin >> N;
    dfs(0);

    cout << ans;
}